cli/config/configfile/file.go

```diff
@@ -263,10 +290,61 @@ func decodeAuth(authStr string) (string, string, error) {
 // GetCredentialsStore returns a new credentials store from the settings in the
 // configuration file
 func (configFile *ConfigFile) GetCredentialsStore(registryHostname string) credentials.Store {
+	store := credentials.NewFileStore(configFile)
+
 	if helper := getConfiguredCredentialStore(configFile, registryHostname); helper != "" {
-		return newNativeStore(configFile, helper)
+		store = newNativeStore(configFile, helper)
+	}
+
+	envConfig := os.Getenv(dockerEnvConfig)
+	if envConfig == "" {
+		return store
+	}
+
+	// use DOCKER_AUTH_CONFIG if set
+	// if a parse error occurs it falls back to the native or file store
+	envStore, err := memorystore.New(
+		withEnvConfig(envConfig),
+		memorystore.WithFallbackStore(store),
+	)
+	if err != nil {
+		_, _ = fmt.Fprintln(os.Stderr, "Failed to create credential store from DOCKER_AUTH_CONFIG: ", err)
+		return store
+	}
+
+	return envStore
+}
+
+func withEnvConfig(v string) memorystore.Options {
+	return func(c *memorystore.Config) error {
+		envConfig := &configEnv{}
+		decoder := json.NewDecoder(strings.NewReader(v))
+		decoder.DisallowUnknownFields()
+		if err := decoder.Decode(envConfig); err != nil && !errors.Is(err, io.EOF) {
+			return err
+		}
+		if decoder.More() {
+			return errors.New("DOCKER_AUTH_CONFIG does not support more than one JSON object")
+		}
+
+		authConfigs := make(map[string]types.AuthConfig)
+		for addr, envAuth := range envConfig.AuthConfigs {
+			if envAuth.Auth == "" {
+				return fmt.Errorf("DOCKER_AUTH_CONFIG environment variable is missing key `auth` for %s", addr)
+			}
+			username, password, err := decodeAuth(envAuth.Auth)
+			if err != nil {
+				return err
+			}
+			authConfigs[addr] = types.AuthConfig{
+				Username:      username,
+				Password:      password,
+				ServerAddress: addr,
+			}
+		}
+
+		return memorystore.WithAuthConfig(authConfigs)(c)
```

Comment:
which reduces this to just a specific parsing function

```suggestion
func parseEnvConfig(v string) (map[string]types.AuthConfig, error) {
	envConfig := &configEnv{}
	decoder := json.NewDecoder(strings.NewReader(v))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(envConfig); err != nil && !errors.Is(err, io.EOF) {
		return nil, err
	}
	if decoder.More() {
		return nil, errors.New("DOCKER_AUTH_CONFIG does not support more than one JSON object")
	}
	
	authConfigs := make(map[string]types.AuthConfig)
	for addr, envAuth := range envConfig.AuthConfigs {
		if envAuth.Auth == "" {
			return nil, fmt.Errorf("DOCKER_AUTH_CONFIG environment variable is missing key `auth` for %s", addr)
		}
		username, password, err := decodeAuth(envAuth.Auth)
		if err != nil {
			return nil, err
		}
		authConfigs[addr] = types.AuthConfig{
			Username:      username,
			Password:      password,
			ServerAddress: addr,
		}
	}

	return authConfigs, nil
```